test/typescript.ts

```diff
@@ -1,33 +1,68 @@
-import {Store, Middleware} from 'redux';
-import thunk, {ThunkAction} from '../index.d.ts';
+import { createStore, applyMiddleware } from 'redux';
+import thunk, { ThunkAction, ThunkMiddleware } from '../index';
 
+type State = {
+  foo: string;
+};
 
-declare const store: Store<{foo: string}>;
+type Actions = { type: 'FOO' } | { type: 'BAR', result: number };
 
-store.dispatch(dispatch => {
-  dispatch({type: 'FOO'});
-});
+type ThunkResult<R> = ThunkAction<R, State, undefined, Actions>;
 
-store.dispatch((dispatch, getState) => {
-  const state = getState();
+const initialState: State = {
+  foo: 'foo'
+};
 
-  const foo: string = state.foo;
-});
+function fakeReducer(state: State = initialState, action: Actions): State {
+  return state;
+}
 
-const middleware: Middleware = thunk.withExtraArgument('bar');
+const store = createStore(fakeReducer, applyMiddleware(thunk as ThunkMiddleware<State, Actions>));
```

Comment:
@Cryrivers I very much appreciate the work you did here!  This is currently working wonderfully in my project 😄 

One issue I had, that I think might be worth documenting somewhere (unless it's just a me thing) is the fact that you cannot type your store to be of type `{ Store } from 'redux'`, otherwise you will get a typescript error when attempting to dispatch thunks


```ts
const store: Store = createStore(/*params*/);    // <-- Typing this to store
store.dispatch((dispatch, getState) => 'foobar') // <-- Causes this to error
```

> _Argument of type 'ThunkAction' is not assignable to parameter of type 'AnyAction'.
>   Property 'type' is missing in type 'ThunkAction'._

Reply:
I guess there's no need to declare `store` as `Store` because `Store` is a generic type. If you use `Store` alone without specifying types, default `dispatch` signature would be used. However, `createStore` returns a `Store` where most generic constraints could be specified, which should be sounder in terms of typing in your case.